content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,114 @@
+
+Here is a plot of the benchmarking numbers:
+
+<div style="margin:0 auto;">
+  <img src="/img/clang-doc-benchmark-numbers.png"><br/>
+</div>
+
+
+### Added Template Mustache HTML Backend
+
+Clang-doc originally used an ad-hoc method of generating HTML, in this project I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however upon further discussion it was decided that the complexity of designing and implementing a new templating language was too much. 
+A LLVM community member suggested I implement a mustache as templating language. 
+Mustache was the ideal templating language, since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 
+In terms of engineering wins, this library was able to cut the direct down on HTML backend significantly dropping 500 lines of code. This library was also designed for general purpose use around LLVM, since there are numerous places in LLVM where various tools generate html in its own way. Using the mustache templating library would be a nice way to standardize the codebase. 
+
+### Improve Clang-Doc HTML Output
+
+The previous version of clang-doc’s output was a pretty minimal bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.
+
+With the new mustache changes an additional backend was added using the specifier (--format=mhtml). That addresses these issues. 
+
+Below is a comparison of the same output between the two backends
+
+
+<div style="margin:0 auto;">
+  <img src="/img/clang-doc-old-html-output.png"><br/>
+</div>
+
+<div style="margin:0 auto;">
+  <img src="/img/clang-doc-new-output.png"><br/>
+</div>
+
+You can also visit the output project on my personal github.io page link
+[here](https://peterchou1.github.io/).
+
+Note: this output is still a work in progress.
+
+## Learning Insight
+
+I've learned a lot in the past few months, thanks to GSOC I now have a much better idea of what it’s like to participate in a large open source project. I received a lot of feedback through PR’s, making RFC and collaborating with other GSOC members. I’d learned a lot about how to interact with the community and solicit feedback. I also learned a lot about instrumentation/profiling code having conducted many experiments in order to try to speed clang-doc up.
```

Comment:
I think it would have been great to detail those experiments a bit more in an earlier section.